src/certificate.rs

```diff
@@ -180,361 +154,172 @@ pub enum PublicKeyInfo {
+    #[asn1(type = "OCTET STRING")]
+    digest: &'a [u8],
 }
 
-/// Digest algorithms.
+/// Creates a new self-signed certificate for the given key. Writes the resulting
+/// certificate to the slot before returning it.
 ///
-/// See RFC 4055 and RFC 8017.
-enum DigestId {
-    /// Secure Hash Algorithm 256 (SHA256)
-    Sha256,
-}
-
-impl x509::AlgorithmIdentifier for DigestId {
-    type AlgorithmOid = &'static [u64];
-
-    fn algorithm(&self) -> Self::AlgorithmOid {
-        match self {
-            // See https://tools.ietf.org/html/rfc4055#section-2.1
-            DigestId::Sha256 => &[2, 16, 840, 1, 101, 3, 4, 2, 1],
-        }
-    }
-
-    fn parameters<W: std::io::Write>(
-        &self,
-        w: cookie_factory::WriteContext<W>,
-    ) -> cookie_factory::GenResult<W> {
-        // Parameters are an explicit NULL
-        // See https://tools.ietf.org/html/rfc8017#appendix-A.2.4
-        x509::der::write::der_null()(w)
-    }
-}
-
-enum SignatureId {
-    /// Public-Key Cryptography Standards (PKCS) #1 version 1.5 signature algorithm with
-    /// Secure Hash Algorithm 256 (SHA256) and Rivest, Shamir and Adleman (RSA) encryption
-    ///
-    /// See RFC 4055 and RFC 8017.
-    Sha256WithRsaEncryption,
-
-    /// Elliptic Curve Digital Signature Algorithm (DSA) coupled with the Secure Hash
-    /// Algorithm 256 (SHA256) algorithm
-    ///
-    /// See RFC 5758.
-    EcdsaWithSha256,
-}
-
-impl x509::AlgorithmIdentifier for SignatureId {
-    type AlgorithmOid = &'static [u64];
+/// `extensions` is optional; if empty, no extensions will be included.
+#[allow(clippy::too_many_arguments)]
+pub fn generate_self_signed<'a>(
+    yubikey: &mut YubiKey,
+    key: SlotId,
+    serial: &[u8],
+    opt_not_after: Option<Time>,
+    subject: &Name<'a>,
+    spkibuf: &[u8],
+    extensions: Option<Extensions<'a>>,
+    alg_id: AlgorithmId,
+) -> Result<Vec<u8>> {
+    let serial_number = UIntBytes::new(serial).unwrap();
+    let spki = match SubjectPublicKeyInfo::from_der(spkibuf) {
+        Ok(spki) => spki,
+        Err(_e) => return Err(Error::ParseError),
+    };
 
-    fn algorithm(&self) -> Self::AlgorithmOid {
-        match self {
-            SignatureId::Sha256WithRsaEncryption => &[1, 2, 840, 113_549, 1, 1, 11],
-            SignatureId::EcdsaWithSha256 => &[1, 2, 840, 10045, 4, 3, 2],
-        }
+    let mut oid = OID_SHA256_WITH_RSA_ENCRYPTION;
+    if OID_EC_PUBLIC_KEY == spki.algorithm.oid {
+        oid = OID_ECDSA_WITH_SHA256;
+    } else if OID_RSA_ENCRYPTION != spki.algorithm.oid {
+        return Err(Error::AlgorithmError);
     }
 
-    fn parameters<W: std::io::Write>(
-        &self,
-        w: cookie_factory::WriteContext<W>,
-    ) -> cookie_factory::GenResult<W> {
-        // No parameters for any SignatureId
-        Ok(w)
-    }
-}
+    let signature = AlgorithmIdentifier {
+        oid,
+        parameters: None,
+    };
 
-/// Certificates
-#[derive(Clone, Debug)]
-pub struct Certificate {
```

Comment:
We should not be removing the owned types and replacing them in the API with `Vec<u8>`; this loses significant type safety. If the issue is that these names conflict with the ones in the `x509` crate, then the latter should be referenced directly as `x509::Certificate` etc. If the issue is that we want to optimise the parsing case, then we should use a `Cow` or similar.

Reply:
The issue was that the Certificate (or really the Decoder below it) maintains a reference to the buffer. I could add a new type that binds the buffer and an Option<Certificate> together and return that.

Reply:
What parts of the certificate borrow? It should probably be an owned data structure

Reply:
Decoder has a ByteSlice: https://github.com/RustCrypto/formats/blob/master/der/src/decoder.rs#L15. ByteSlice has a reference here: https://github.com/RustCrypto/formats/blob/master/der/src/byte_slice.rs#L16.

Reply:
Yes, but once the certificate is decoded everything should be owned and no longer reference the input buffer.

Where does the lifetime of borrowed data come into play?

Reply:
All of the buffers in the structure reference points in the original buffer. So this code snip:

```
        let buf = <read buffer>;
        let cert =  Certificate::from_der(buf.as_slice());
```

Yields these addresses in a sample run just now:

0x00007fa5cb009200 - buffer (1281 bytes total)
0x00007fa5cb009337 - spki
0x00007fa5cb009459 - first extension
0x00007fa5cb009601 - signature buffer

Reply:
Aah yeah, an owned form of SPKI seems needed.

There's the [`PublicKeyDocument`](https://docs.rs/spki/latest/spki/struct.PublicKeyDocument.html) document type which we could potentially make work for this application. I think it's kind of like what you're describing where it's a newtype for a `Vec<u8>` which impls the [`der::Document`](https://docs.rs/spki/latest/spki/struct.PublicKeyDocument.html#impl-Document%3C%27a%3E) trait for decoding.

Reply:
I will give a look at preparing a CertificateDocument class and return that instead of the buffer in the spots where I moved from struct to buffer. I'll put that in x509 and move the OIDs and nametostr while at it.

Reply:
> We should not be removing the owned types and replacing them in the API with Vec<u8>; this loses significant type safety

There's a "document" type convention for this in the other https://github.com/RustCrypto/formats/ and this PR adds a `CertificateDocument` type: https://github.com/RustCrypto/formats/pull/393

These types are effectively a newtype wrapper around a `Vec<u8>` which is guaranteed to parse successfully as a particular type, i.e. a `Certificate` in this case. They impl the [`der::Document`](https://docs.rs/der/latest/der/trait.Document.html) trait which handles serializing/deserializing DER-encoded documents including PEM (de)serialization as well as reading from and writing to DER or PEM-encoded files.